File: python/alpm_qo.py
import alpm

what = "/usr/bin/pacman"

alpm.initialize("/")
db = alpm.db_register('local')

i = alpm.db_getpkgcache(db)
while i:
	pkg = alpm.void_to_PM_PKG(alpm.list_getdata(i))
	j = alpm.void_to_PM_LIST(alpm.pkg_getinfo(pkg, alpm.PKG_FILES))
	while j:
		if alpm.void_to_char(alpm.list_getdata(j)) == what[1:]:
			print alpm.void_to_char(alpm.pkg_getinfo(pkg, alpm.PKG_NAME))
		j = alpm.list_next(j)
	i = alpm.list_next(i)

File: perl/alpm_qo.pl
use Alpm::Core;

sub getowner
{
	my $what = shift;
	$i = Alpm::Core::alpm_db_getpkgcache($db);
	while($i)
	{
		$pkg = Alpm::Core::void_to_PM_PKG(Alpm::Core::alpm_list_getdata($i));
		$j = Alpm::Core::void_to_PM_LIST(Alpm::Core::alpm_pkg_getinfo($pkg, $Alpm::Core::PM_PKG_FILES));
		while($j)
		{
			if(Alpm::Core::void_to_char(Alpm::Core::alpm_list_getdata($j)) eq $what)
			{
				return Alpm::Core::void_to_char(Alpm::Core::alpm_pkg_getinfo($pkg, $Alpm::Core::PM_PKG_NAME));
			}
			$j = Alpm::Core::alpm_list_next($j);
		}
		$i = Alpm::Core::alpm_list_next($i);
	}
}

Alpm::Core::alpm_initialize("/");
$db = Alpm::Core::alpm_db_register('local');

print getowner("usr/bin/pacman") . "\n";
print getowner("lib/libc.so.6") . "\n";
